Checksum verification: add content (#9)
3mdeb/verified-boot: verified_boot_main.md

### Checksum verification

Checksum as a concept, is a redundancy check for detecting variations in data.
Checksums are created as result of calculating binary values of a block or

> packet of data using sophisticated algorithms. Comparing checksums allows for

Reviewer: > sophisticated algorithms

parity checksum == xor all bits
It can use sophisticated algorithms but there are also very simple ones.

> verifying data integrity, yet is truthfulness is dependent of algorithm

Reviewer: ```suggestion
verifying data integrity, yet its truthfulness is dependent of algorithm
```

used[^1].

Checksums values come in form of unique, fixed string of characters referred to
as hash. Exact form of hash is dependent on algorithm used. Checksum algorithms
ensure even tiniest changes made to a block of data, would result in checksum
value changing completely. This property makes them easy to compare without
need for any additional components. Common algorithms include: sha-256, md5 or

> sha-1[^2].

Reviewer: > even tiniest changes made to a block of data, would result in checksum
value changing completely. This property

I think that's property of good cryptographic hash function (I think my professor called it mixing property i.e. small input change results in big output change).

> Checksums values come in form of unique

I'm not sure what you mean here. It's impossible for checksums (or hashes) to be unique unless space of input values is <= space of output values.

> fixed string of characters
Checksum can be integer value e.g. IPv4 header checksum which is sum of all 2-byte values in header (though I guess you could view them as string characters (unless they are unprintable))

#### Checksum verification in chain of trust context

> Checksum verification can be used as a part of chain of trust mechanism.

Reviewer: I think any chain of trust would require use of cryptographic hash functions (which have more stringent properties) instead of any checksum.

Particular example is Google's Verified Boot found on Android based devices.
Verified boot cryptographically verifies executable code and data as the device

> boots. It does so for both files and partitions using multiple mechanisms, one

Reviewer: ```suggestion
boots. It does so for both files and partitions using multiple mechanisms,
```

one of them being [dm-verity](#dm-verity). In Verify Boot implementation hashes
are stored on dedicated partitions and are signed by the root of trust[^3].

#### debsums

A supplementary integrity checks can also be performed on OS level.

> `debsums` is an utility for verifying debian package files against their

Reviewer: ```suggestion
`debsums` is a utility for verifying Debian package files against their
```

checksums. It is primarily intended for determining if installed filed were
modified or damaged. It also serves limited usage as a security tool[^4].

#### References

> [^1] [Checksum Definition](https://www.linfo.org/checksum.html)
> [^2] [Exploring Checksums: Types, Uses, And Verification Methods](https://linuxsecurity.com/features/what-are-checksums-why-should-you-be-using-them)
> [^3] [Verify Boot](https://source.android.com/docs/security/features/verifiedboot/verified-boot)
> [^4] [debsums](https://manpages.ubuntu.com/manpages/trusty/man1/debsums.1.html)

Reviewer: ```suggestion
[^1]: [Checksum Definition](https://www.linfo.org/checksum.html)
[^2]: [Exploring Checksums: Types, Uses, And Verification Methods](https://linuxsecurity.com/features/what-are-checksums-why-should-you-be-using-them)
[^3]: [Verify Boot](https://source.android.com/docs/security/features/verifiedboot/verified-boot)
[^4]: [debsums](https://manpages.ubuntu.com/manpages/trusty/man1/debsums.1.html)
```

### Checksum verification

Checksum is a redundancy check for detecting variations in data or data
streams. Checksums are created as result of calculating binary values of a
block or packet of data using dedicated algorithms. Comparing checksums allows
for verifying data integrity, yet it's truthfulness is dependent of algorithm
used[^1].

#### Checksums and hash codes

Transforming a portion of data of any size, into a code with predefined size is
referred to as hashing process. A result of performing a hash function is
output in form of a hash code (or simply hash). Hash codes are irreversible by
definition, in theory a hash cannot be used to recreate source data.

Checksums are based on hash codes concept, yet they serve different purposes.
Checksum algorithms are optimized for calculation speed in expense of security.
Checksums are focused on checking data integrity, serving as a digital
fingerprints. A hash codes and cryptographic functions are much broader concept
that focuses on security and authentication[^2].

#### Common hashing algorithms

Common hashing algorithms include[^3]:

> * MD-5 - created in 1991, but now considered compromised as it was discovered
> how to decode it.

Reviewer 2: I don't think "decode" is the right term here, since MD5 isn't a reversible encryption algorithm. What has been discovered are successful collision attacks against MD5, so maybe this would be more accurate:

```suggestion
* MD-5 - created in 1991, but now considered compromised as successful collision attacks
against it have been demonstrated.
```

https://web.archive.org/web/20170517115509/http://www.win.tue.nl/hashclash/On%20Collisions%20for%20MD5%20-%20M.M.J.%20Stevens.pdf and https://web.archive.org/web/20170325033522/http://www.win.tue.nl/hashclash/rogue-ca/ can be used as references for this.

Author: Fixed.

* SHA family - developed by US government. A recent SHA-3 version uses
randomness in the code making it more difficult to decode. Currently an

> industry standard.

Reviewer 2: What does "randomness in the code" mean? This would probably benefit from a reference. Similarly, "decode" may need replaced with "find collisions for".

Author: Fixed and added reference.

Hashing algorithm produce hash values, yet they are often used as a way to
verify data integrity, serving a role of a checksum. This is why it is not
uncommon to see an output of hashing algorithms being referred to as checksum.

#### Hash verification in chain of trust context

Hashing functions can be used as a part of chain of trust mechanism.
Particular example is Google's Verified Boot found on Android based devices.
Verified boot cryptographically verifies executable code and data as the device
boots. It does so for both files and partitions using multiple mechanisms, one
of them being [dm-verity](#dm-verity). In Verify Boot implementation hashes
are stored on dedicated partitions and are signed by the root of trust[^3].

#### Extending chain of trust to OS

> A supplementary integrity checks can also be performed on OS level.

Reviewer 3: Good to insert newline here?

`debsums` is a utility for verifying Debian package files against their
checksums. The MD-5 algorithm is used in the process. `debsums` is primarily
intended for determining if installed files were modified or damaged. It also
serves limited usage as a security tool[^5].

> 

Reviewer 3: debsums isn't great of this yet indeed. But that should not be understood as discouraging.

Worth mentioning other tools? Worth mentioning that this is not a inherent issue and fixable in theory?

Mention debcheckroot does exist and that other similar tools might exist?

Author: >Worth mentioning other tools? 

Sure.

>Worth mentioning that this is not a inherent issue and fixable in theory?

IMO, we should focus on the state it is in currently. My take is that wiki should not "wonder about the future".

> Mention debcheckroot does exist and that other similar tools might exist

Ack.

Author: Added `debcheckroot` and `AIDE`.
